File: Windows/Entities/Deck.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Cribbage.Entities
{
    public class Deck
    {  
        public List<Card> Cards { get; set; }
        public Card CutCard { get; set; }

        //static Random _random = new Random();
        //public Deck()
        //{
        //    Reset();
        //}

        //public void Reset()
        //{
        //    _cards.Clear();
        //    foreach (Suit suit in Enum.GetValues(typeof(Suit)))
        //    {
        //        for (int i = 1; i < 14; i++)
        //            _cards.Add(new Card(suit, (CardValue)i));
        //    }
        //}

        //public void Shuffle()
        //{
        //    //http://en.wikipedia.org/wiki/Fisher-Yates_shuffle
        //    for (int i = _cards.Count - 1; i > 0; i--)
        //    {
        //        int k = _random.Next(i + 1);
        //        Card temp = _cards[i];
        //        _cards[i] = _cards[k];
        //        _cards[k] = temp;
        //    }
        //}

        //public Card[] Deal(int count)
        //{
        //    if (count >= _cards.Count)
        //    {
        //        try
        //        {
        //            Card[] temp = new Card[count];
        //            _cards.CopyTo(0, temp, 0, count);
        //            _cards.RemoveRange(0, count);

        //            return temp;
        //        }
        //        catch
        //        {
        //            return null;
        //        }
        //    }

        //    return null;
        //}

        //public void Cut(int index)
        //{
        //    Card[] temp = new Card[_cards.Count];

        //    _cards.CopyTo(index, temp, 0, _cards.Count - index);
        //    _cards.CopyTo(0, temp, _cards.Count - index, index);
        //    _cards.Clear();
        //    _cards = temp.ToList();

        //    _cutCard = _cards[0];
        //    _cards.RemoveAt(0);
        //}
    }
}

File: server/cribbage/deck.ts
import {ICard} from '../cribbage';

export class Deck {
    private _cards: Array<ICard>;
    private _cutCard: ICard;

    get topCard(): ICard {
        return this._cards.length > 0 ? this._cards[0] : null;
    }
    get cardIds(): string {
        var s = "";
        this._cards.forEach(function(card, index, array){
            if (index > 0) {
                s += ",";
            }
            s += card.id;
        });
        return s;
    }
    get cutCard(): ICard {
        return this._cutCard;
    }

    constructor(cards?: Array<ICard>) {
        if (cards) {
            this._cards = cards;
            //this.shuffle();
        }
    }

    public deal(): ICard {
        return (this._cards.length > 0) ? this._cards.shift() : null;
    }
    public shuffle(): void {
        //http://en.wikipedia.org/wiki/Fisher-Yates_shuffle
        for (var i = this._cards.length-1; i>0; i--) {
            var j = getRandomInt(0, i+1);
            var temp:ICard = this._cards[i];
            this._cards[i] = this._cards[j];
            this._cards[j] = temp;
        }

        function getRandomInt(min:number, max:number): number {
            min = Math.ceil(min);
            max = Math.floor(max);
            return Math.floor(Math.random() * (max - min)) + min;
        }
    }
    public cut(index: number): ICard {
        var bottom = this._cards.slice(index);
        for (var i=0; i<index; i++) {
            bottom.push(this._cards[i]);
        }
        this._cards = bottom;

        return this.deal();
    }
}